src/parser/jobs/blm/modules/ABC.js

```diff
@@ -0,0 +1,125 @@
+const DURATION_ERROR_OFFSET = 1000
+
+export default class AlwaysBeCasting extends Module {
+	static handle = 'blackmage_abc'
+	static i18n_id = i18nMark('blm.abc.title')
+	static title = 'Times you did literally nothing'
+	static displayOrder = DISPLAY_ORDER.ABC
+
+	static dependencies = [
+		'timeline',
+		'gcd',
+		'invuln',
+	]
+
+	_noCastWindows = {
+		current: null,
+		history: [],
+	}
+
+	constructor(...args) {
+		super(...args)
+
+		this.addHook('begincast', {by: 'player'}, this._onBegin)
+		this.addHook('cast', {by: 'player'}, this._onCast)
+		this.addHook('death', {to: 'player'}, this._onDeath)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCast(event) {
+		//better than using 2.5s I guess
+		const gcdLength = this.gcd.getEstimate(false)
+		//don't check the time that you actually spent casting
+		if (!this._noCastWindows.current) {
+			this._noCastWindows.current = {
+				start: event.timestamp,
+			}
+			return
+		}
+		//check if it's been more than a gcd length
+		if (event.timestamp - this._noCastWindows.current.start > gcdLength + GCD_ERROR_OFFSET) {
+			this._stopAndSave(event.timestamp)
+		}
+		//this cast is our new last cast
+		this._noCastWindows.current = {
+			start: event.timestamp,
+		}
+	}
+
+	_onBegin(event) {
+		const gcdLength = this.gcd.getEstimate(false)
+		if (this._noCastWindows.current) {
+			if (event.timestamp - this._noCastWindows.current.start > gcdLength + GCD_ERROR_OFFSET) {
+				this._stopAndSave(event.timestamp)
+			}
+			this._noCastWindows.current = null
+		}
+	}
+
+	//reset to not count the time you lie on the ground as time you aren't casting : ^)
+	_onDeath() { this._noCastWindows.current = null }
+
+	_stopAndSave(endTime = this.parser.currentTimestamp) {
+		const tracker = this._noCastWindows
+
+		// Already closed, nothing to do here
+		if (!tracker.current) {
+			return
+		}
+
+		// Close the window
+		tracker.current.stop = endTime
+		tracker.history.push(tracker.current)
+		tracker.current = null
+	}
+
+	_onComplete(event) {
+		//finish up
+		this._stopAndSave(event.timestamp)
+		//filter out invuln periods
+		this._noCastWindows.history = this._noCastWindows.history.filter(windows => {
+			return this.invuln.getInvulns('all', windows.start, windows.stop).length === 0
+		})
+		//filter out negative durations
+		this._noCastWindows.history = this._noCastWindows.history.filter(windows => windows.stop - windows.start > DURATION_ERROR_OFFSET)
+	}
+
+	output() {
+		if (this._noCastWindows.history.length > 0) {
+			return <Table collapsing unstackable compact="very">
+				<Table.Header>
+					<Table.Row>
+						<Table.HeaderCell><Trans id="blm.abc.timestamp-header">Timestamp</Trans></Table.HeaderCell>
+						<Table.HeaderCell><Trans id="blm.abc.duration-header">Duration</Trans></Table.HeaderCell>
+						<Table.HeaderCell></Table.HeaderCell>
+					</Table.Row>
+				</Table.Header>
+				<Table.Body>
+					{this._noCastWindows.history.map(notCasting => {
```

Comment:
As with the other PR that's open at the moment - I'm really wary of how much this output could inflate. A good player may have sub-5 pops of this... but I've seen more than my fair share of BLMs who stop to pick their nose between each GCD. It may well easily inflate to more than a pageheight.

While I won't hard no this, It's worth discussing resolutions to this issue - as it _would_ be nice to resolve it in some manner.

cc @AzariahAeroborn

To me, this _feels_ like a timeline sorta shtick. But we all know how much the timeline sucks ass. There's already been some discussion about it in #caster_development, worth continuing it.

Reply:
Keep in mind the module is very lenient. I don't underestimate our users being bad but the module looks at the time between 2 actions (any actions) and sees if there has been gcd+0.1s inbetween. Any cast, even a canceled one or ogcd or sprint counts here as a try and resets the timer. My aim was that people preferably never see the output and if you do you really should wake the fuck up

Reply:
Maybe I explained that wrong. Since the module doesn't know if your last cast was instant or not it assumes a gcd ticking down after your last action and then expects another action a gcd later. If that doesn't happen the module triggers. So you need to basically do nothing for ~2.5-5s for it to trigger